libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProvider.kt

```diff
@@ -0,0 +1,249 @@
+import com.mapbox.mapboxsdk.style.expressions.Expression.exponential
+import com.mapbox.mapboxsdk.style.expressions.Expression.get
+import com.mapbox.mapboxsdk.style.expressions.Expression.interpolate
+import com.mapbox.mapboxsdk.style.expressions.Expression.literal
+import com.mapbox.mapboxsdk.style.expressions.Expression.match
+import com.mapbox.mapboxsdk.style.expressions.Expression.product
+import com.mapbox.mapboxsdk.style.expressions.Expression.stop
+import com.mapbox.mapboxsdk.style.expressions.Expression.switchCase
+import com.mapbox.mapboxsdk.style.expressions.Expression.zoom
+import com.mapbox.mapboxsdk.style.layers.LineLayer
+import com.mapbox.mapboxsdk.style.layers.Property
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconAllowOverlap
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconIgnorePlacement
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconImage
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconPitchAlignment
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconSize
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineCap
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineColor
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineJoin
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.lineWidth
+import com.mapbox.mapboxsdk.style.layers.SymbolLayer
+import com.mapbox.navigation.ui.internal.route.RouteConstants.ALTERNATIVE_ROUTE_LAYER_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.ALTERNATIVE_ROUTE_SHIELD_LAYER_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.ALTERNATIVE_ROUTE_SOURCE_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.DEFAULT_ROUTE_DESCRIPTOR_PLACEHOLDER
+import com.mapbox.navigation.ui.internal.route.RouteConstants.DESTINATION_MARKER_NAME
+import com.mapbox.navigation.ui.internal.route.RouteConstants.ORIGIN_MARKER_NAME
+import com.mapbox.navigation.ui.internal.route.RouteConstants.PRIMARY_ROUTE_LAYER_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.PRIMARY_ROUTE_SHIELD_LAYER_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.PRIMARY_ROUTE_SOURCE_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.PRIMARY_ROUTE_TRAFFIC_LAYER_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.PRIMARY_ROUTE_TRAFFIC_SOURCE_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.WAYPOINT_DESTINATION_VALUE
+import com.mapbox.navigation.ui.internal.route.RouteConstants.WAYPOINT_LAYER_ID
+import com.mapbox.navigation.ui.internal.route.RouteConstants.WAYPOINT_ORIGIN_VALUE
+import com.mapbox.navigation.ui.internal.route.RouteConstants.WAYPOINT_PROPERTY_KEY
+import com.mapbox.navigation.ui.internal.route.RouteConstants.WAYPOINT_SOURCE_ID
+import com.mapbox.navigation.ui.internal.utils.MapImageUtils
+import com.mapbox.navigation.ui.route.RouteStyleDescriptor
+
+interface MapboxRouteLayerProvider : RouteLayerProvider {
```

Comment:
This implementation doesn't have or need any mutable state. All of the functions are pure. Putting the implementation in an interface enforces the no mutable state concept.  It prevents the casual adding of mutable state. In other classes in the code base I've seen cases of un-necessary mutable state which have led to unpredictable side effects and thus bugs that could have been prevented. While increasing the "safety" of this implementation there's still flexibility for future modification. If there's a necessity to add mutable state to this in the future it can be changed to a class and the change won't effect consumers of this implementation since it's implementing the RouteLayerProvider which defines the contract.  Also there is a factory to instantiate this interface so if for some reason it needed to get changed to a class in the future only the factory would require a code change since consumers don't instantiate MapboxRouteLayerProvider directly. Changing this to a class in the future, if needed, won't break SEMVER since it still implements the contract interface and the consumers interact with the factory to get an instance.